syntax = "proto3";

package thresholdsignature;

message TimeRequest{
    string name=1;
    string uuid=2;
}

message TimeResponse{
    string timestring=1;
}

message player{
    string name=1;
    string uri=2;
    string addr=3;
    string port=4;
}



message ThresholdGroup{
    string grpid=1;
    string proposer=2;
    int32 m=3;
    int32 n=4;
    int32 t=5;
    
    message GroupMember{
        string userid=1;
        int32 ordinal=2;
        string useruri=3;
        string userip=4;
        string userport=5;
    }
    
    repeated GroupMember member=6; 
}

message ThresholdGroupResponse{
    string groupid=1;
    string userid=2;
    bool userset=3;
}

message RegisterPlayerRequest{
    repeated player playerID=1;
}

message RegisterPlayerResponse{
    bool success=1;
}

message GetPlayerListRequest{ 
    string name=1;
}

message GetPlayerListResponse{
    repeated player playerID=1;
}

message CreateThresholdGroup{
    string proposerID=1;
    int32 recombinationThreshold=2;
    int32 maxnumberofplayers=3;
}

message CreateThresholdGroupResponse{
    bool groupset=1;
    string grpid=2;
}

message InviteToGroupRequest{
    string proposerUserID=1;
    string groupid=2;
}

message InviteToGroupResponse{
    string userid = 1;
    string groupid =2;
    string useruri=3;
    string userip=4;
    string userport=5;
    bool acceptInvite=6;
}

message BroadCastGroupDetails{
    string groupid=1;
    repeated ThresholdGroup grpMetaData=2;
}

message BroadCastGroupDetailsResponse{
    string groupid=1;
    string userid = 2;
    bool userset= 3;
}

message DeleteGroupRequest{
    string userid=1;
    string groupid=2;
}

message DeleteGroupResonse{
    string userid=1;
    string groupid=2;
    bool groupdeleted=3;
}

message DeleteGroupPlayerRequest{
    string userid=1;
    string groupid=2;
}

message DeleteGroupPlayerResponse{
    string userid=1;
    string groupid=2;
    bool deleted=3;
}

message SecretSharingRequest{
    string userid=1;
    string groupid=2;
    string calctype=3;
}

message SecretSharingResponse{
    string userid=1;
    string groupid=2;
    string calctype=3;
    bool sharedsecret=4;
}

message InitEvalSharingRequest{
    string userid=1;
    string groupid=2;
    string calctype=3;
}

message InitEvalSharingResponse{
    string userid=1;
    string groupid=2;
    string calctype=3;
    bool evalsshared=4;
}

message EvalSharingRequest{
    string userid=1;
    string groupid=2;
    string fromord=3;
    string eval=4;
}

message EvalSharingResponse{
    string userid=1;
    string groupid=2;
    bool evalreceived=3;
}

message SecretSharingPlayerDataRequest{
    string userid=1;
    string groupid=2;
    string calctype=3;
}

message singleString{
    string val=1;
}

message stringPair{
    string ordinal=1;
    string eval=2;
}

message PublicEvals{
    string ordinal=1; 
    repeated stringPair evals=2;
}

message SinglePublicEval{
    string ordinal=1;
    string eval=2;
}

message OrdinalVW{
    string ordinal=1;
    string v=2;
    string w=3;
}

message HiddenPoly{
    repeated singleString val=1;
}

message KeyStringAndValueListStringPair{
    string key=1;
    repeated stringPair value=2;
}

message KeyStringAndValueListString{
    string key=1;
    repeated string value=2;
}

message KeyStringAndValueListRepeatedString{
    string key=1;
    repeated singleString value=2;
}

message SecretSharingPlayerDataResponse{
    string userid=1;
    string groupid=2;
    string calctype=3;
    
  
    int32 ordinal=4;
    repeated KeyStringAndValueListStringPair hiddenevals = 7;
    repeated HiddenPoly hiddenpoly=8;
    
}

message SecretSharingCollatedPlayerRequest{
    // This will have all the data from message ID-20 from all clients
    string userid=1;
    string groupid=2;
    string calctype=3;
    repeated KeyStringAndValueListRepeatedString collatedhiddenpolys=5;
    repeated KeyStringAndValueListStringPair collatedhiddenevals=6;
    
}

message SecretSharingCollatedPlayerResponse{
    // This message is data validated by all players in the group
    // the final message of
    string userid=1;
    string groupid=2;
    string calctype=3;
    bool collatedplayerdatavalidated=4;
    
}


message EphemeralKeyDataRequest{
    string userid=1;
    string groupid=2;
}


message EphemeralKeyDataResponse{
    string userid=1;
    string groupid=2;
    bool emphemeralkeydataset=3;
}

message EphemeralKeyPlayerDataRequest{
    string userid=1;
    string groupid=2;
}


message EphemeralKeyPlayerDataResponse{
    string userid=1;
    string groupid=2;
    int32 ordinal=3;
    string kalpha=4;
    string alphag=5;
}


message AllEphemeralKeyPlayerRequest{
    string groupid=1;
    repeated OrdinalVW collatedvws=2;
}


message AllEphemeralKeyPlayerResponse{
    string userid=1;
    string groupid=2;
    string ordinal=3;
    bool ephemeralkeyplayerset=4;
}


message GenerateSignatureRequest{
    string userid=1;
    string groupid=2;
    string msg=3;
    int32 emphemeralkeyindex=4;
}


message GenerateSignatureResponse{
    string userid=1;
    string groupid=2;
    string msg=3;
    int32 emphemeralkeyindex=4;
    repeated stringPair partialsignatures=5;

}

 
message SignaturePlayerDataRequest{
    string userid=1;
    string groupid=2;
    string msg=3;
    int32 emphemeralkeyindex=4;
}

message SignaturePlayerDataResponse{
    string userid=1;
    string groupid=2;
    string ordinal=3;
    string signatureshare=4;
    int32 emphemeralkeyindex=5;
}


message GenerateGroupPrivateKeyRequest{
    string userid=1;
    string groupid=2;
    repeated singleString players=3; 
}

message GenerateGroupPrivateKeyResponse{
    string userid=1;
    string groupid=2; 
    bool privatedatashared=3;
}

message PlayerPrivateShareExchangeRequest{
    string userid=1;
    string groupid=2;
}

message PlayerPrivateShareExchangeResponse{
    string userid=1;
    string groupid=2;
    bool privatedatashared=3;
}

message PlayerPrivateKeyShareRequest{
    string userid=1;
    string groupid=2;
    string ordinal=3;
    string privatekeyshare=4;
}

message PlayerPrivateKeyShareResponse{
    string userid=1;
    string groupid=2;
    bool sharedprivatekeyshare=3;
}
